recipes/full_finetune_single_device_perf_tools_showcase.py

```diff
@@ -0,0 +1,400 @@
+                        == self.max_steps_per_epoch
+                    ):
+                        break
+
+                    # Send the signal to the profiler that the next step has started
+                    self._perf_profiler.step()
+
+                    input_ids, labels = batch
+                    input_ids = input_ids.to(self._device)
+                    labels = labels.to(self._device)
+                    logits = self._model(input_ids)
+                    # Shift so that tokens < n predict n
+                    logits = logits[..., :-1, :].contiguous()
+                    labels = labels[..., 1:].contiguous()
+                    logits = logits.transpose(1, 2)
+                    # Compute loss
+                    loss = self._loss_fn(logits, labels)
+                    # Note: We're always logging the loss before normalizing it
+                    # Check if this is the norm or not
+                    pbar.set_description(f"{curr_epoch+1}|{idx+1}|Loss: {loss.item()}")
+                    if self.total_training_steps % self._log_every_n_steps == 0:
+                        self._metric_logger.log_dict(
+                            {
+                                "loss": loss.item(),
+                                "lr": self._optimizer.param_groups[0]["lr"],
+                                "gpu_resources": torch.cuda.memory_allocated(),
+                            },
+                            step=self.total_training_steps,
+                        )
+
+                    loss = loss / self._gradient_accumulation_steps
+                    loss.backward()
+                    if self._should_update_weights(idx):
+                        self._optimizer.step()
+                        self._optimizer.zero_grad(set_to_none=True)
+
+                        # Update the number of steps when the weights are updated
+                        self.total_training_steps += 1
+
+                    # Log peak memory for iteration
+                    if (
+                        self.total_training_steps % self._log_peak_memory_every_n_steps
+                        == 0
+                    ):
+                        log.info(
+                            utils.memory_stats_log("Memory Stats:", device=self._device)
+                        )
+            self.epochs_run += 1
+            self.save_checkpoint(epoch=curr_epoch)
+
+    def cleanup(self) -> None:
+        self._metric_logger.close()
+
+
+@config.parse
+def recipe_main(cfg: DictConfig) -> None:
+    """
+    Entry point for the recipe.
+
+    Configurable parameters are read in the following order:
+        - Parameters specified in config (see available configs through ``tune ls``)
+        - Overwritten by arguments from the command-line
+    """
+    recipe = FullFinetuneRecipeSingleDevicePerfToolsShowcase(cfg=cfg)
+    recipe.setup(cfg=cfg)
+    recipe.train()
+    recipe.cleanup()
+
+
+if __name__ == "__main__":
+    sys.exit(recipe_main())
```

Comment:
I don't think we need an entirely separate recipe + config just for the profiling component. I think one idea is to build the profiler as a component, and integrate it into a single _existing_ recipe as a showcase. Rest of the recipes can have it off.

Reply:
Thanks for raising this @rohan-varma! I actually thought hard about what is the best way to do this as well. My concern around integrating it with 1 existing recipe is that for hobbyists that won't look deep in the recipe, integrating will cause confusion to them as profiler is not necessary for every training and mainly for debugging purpose. Having a separate recipe is more clear although it will increase the complexity of our recipe system by introducing 1 more recipe. 

But I'm open to integrate the component into an existing recipe. Looking forward to other people's thought.